File: scripts/simple_number_sub_pub_srv.py
#!/usr/bin/env python3
"""
    ROS node doing the following
	- Subscribes to simple_number_publisher
	- Publishes the sum of received numbers
	- Provide a service to reset the count 
    
    Author: Jari Honkanen
"""

import rospy
from std_msgs.msg import Int64
from std_srvs.srv import SetBool

sum = 0
pub = None

def handle_callback_number(msg):
	global sum
	sum += msg.data
	new_msg = Int64()
	new_msg.data = sum
	pub.publish(new_msg)

def handle_callback_reset_adder(request):
	
	if request.data:
		global sum
		sum = 0
		return True, "[INFO] Counter reset successful"

	return False, "[ERROR] Counter reset failed"

if __name__ == '__main__':

	rospy.init_node("simple_number_sub_pub_srv")

	sub = rospy.Subscriber("/number", Int64, handle_callback_number)

	pub = rospy.Publisher("/number_sum", Int64, queue_size=12)

	service = rospy.Service("/reset_adder", SetBool, handle_callback_reset_adder)

	rospy.loginfo("[INFO] simple_number_sub_pub_srv started")

	rospy.spin()

File: my_simple_robot_nodes/src/simple_number_sub_pub_srv.cpp
/*-------------------------------------------------------------------*\
  NAME
    simple_number_publisher.cpp

  DESCRIPTION
    ROS node doing the following:
	- Subscribes to simple_number_publisher
	- Publishes the sum of received numbers
	- Provide a service to reset the count 

  AUTHOR
    Jari Honkanen
\*-------------------------------------------------------------------*/

#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <std_srvs/SetBool.h>

int sum = 0;
ros::Publisher pub;

void handle_callback_number(const std_msgs::Int64& msg)
{
	sum += msg.data;
	std_msgs::Int64 new_msg;
	new_msg.data = sum;
	pub.publish(new_msg);
}

bool handle_callback_reset_adder(std_srvs::SetBool::Request &req, std_srvs::SetBool::Response &res)
{
	if (req.data) {
		sum = 0;
		res.message = "[INFO] Counter reset successful";
		res.success = true;
	}
	else {
		res.message = "[ERROR] Counter reset failed";
		res.success = false;
	}

	return true;
}


int main (int argc, char **argv)
{
	ros::init(argc, argv, "number_counter");
	ros::NodeHandle nh;

	ros::Subscriber sub = nh.subscribe("/number", 1000, handle_callback_number);

	pub = nh.advertise<std_msgs::Int64>("/number_sum", 12);

	ros::ServiceServer reset_service = nh.advertiseService("/reset_adder", handle_callback_reset_adder);

	ROS_INFO("[INFO] simple_number_sub_pub_srv started");

	ros::spin();
}
